### ISSUE TITLE
bug in "event" export?

### ISSUE THREAD

**Issue by ChiaraPalladino** [pelagios/recogito2#465]
I just realized that event annotations are not exported with the event type in the csv format. Whereas we have person and place in the type field, event does not appear.

**Comment by rsimon** [comment]
Ah, well spotted! Do you need this fixed in the production instance right now? I fixed it in the codebase. But won't be doing any updates to the live instance before Mon 12. (Holidays, plus that huge combined platform-upgrade-office-location-move thing.) If that's sufficient, it would save some work :-)

**Comment by ChiaraPalladino** [comment]
Actually I can solve the problem manually, I don't have that many annotations. It also seems that the exported data mix old and new annotations in a weird way - I spotted several annotations that I deleted on Recogito, which still survived in the export file.

**Comment by rsimon** [comment]
Hm, that's weird though... Can you share the document and point me at the annotations in question?

**Comment by rsimon** [comment]
Rather than deleted annotations, could it be that those are annotations that still exist, but are "underneath" a newer annotation? (Same start and end in the text?)

That's a different issue which I think we briefly discussed already. Will be addressed soon, too.

**Comment by ChiaraPalladino** [comment]
That was my first thought, but some of them are not stratified. I deleted them and tried a new export. 

I am going to share the document with you via Recogito.

**Comment by ChiaraPalladino** [comment]
http://recogito.pelagios.org/document/vjar6k90gzwsel (I just noticed I already shared the document with you!).

**Comment by rsimon** [comment]
Thanks, can you point me to (the IDs of) some annotations that should be deleted, but appear in the CSV?

**Comment by ChiaraPalladino** [comment]
wow, it's actually **not saving** some of my annotations even in the annotation interface. this may need some more systematic testing, i'm confused now.

**Comment by rsimon** [comment]
Hm, is it showing the error message in the header bar? I can't seem to see any error on the server at the moment, so must be something client-side.

**Comment by ChiaraPalladino** [comment]
it doesn't. that happened yesterday a couple of times, but i just relogged in that case and it was fine, although it was surprisingly frequent. now I'm logged in, and although it shows me the time of the most recent annotation ("one minute ago"), it does not appear in the document.

**Comment by rsimon** [comment]
I just made a comment (and deleted again), so the "one minute ago" in the index page is probably from me. Otherwise, it doesn't look as if any annotations got stored on the server in the last 29 minutes?

**Comment by ChiaraPalladino** [comment]
which is very strange, because I was making annotations since two minutes ago. "one minute ago" was just an example :)

**Comment by rsimon** [comment]
Hm, looks like there are new annotation now. Did you change anything?

**Comment by ChiaraPalladino** [comment]
yes, I just tested again. 

I think I see what is weird. I need to do more checks on the document, but for some reason the annotations are stored, _but_ not visualized in recogito! Let me check this more thoroughly and I'll get back at you today.

**Comment by ChiaraPalladino** [comment]
![image](https://user-images.githubusercontent.com/15249889/35733325-3b5d29b6-081d-11e8-8609-329345a58e8e.png)

Example here: the string τὸν ἕως Κασπίας ἰσθμόν was annotated yesterday as an event. it _is_ in fact saved, and it appears in the exported file. however, it does not appear visible in the interface. which may be the problem that brought to multiple annotations storage as well (because if I don't see an annotation, I believe I missed that textchunk and I repeat that annotation again).

**Comment by rsimon** [comment]
Thanks. I need to trace this down more thoroughly... but here's one observation I made just now:

If you have stratified annotations, loading them on a fresh page seems to render the topmost annotation __only__. 

I.e. assume you load the document, and then delete the top-most of a bunch of stratified annotations. In the UI, the highlight will be removed - as if there are no more annotations left. However, those stratified annotations are still stored. That also means that if you now create a new annotation on the (now) empty space, you'll just add another annotation on top of those that already exist.

To get firmly rid of the stratified ones, at the moment you'd need to delete the topmost (highlight will be gone). Then reload the page (highlight will be back, with the next top-most annotation in the stratigraphy.). Delete that highlight - and repeat.

Not 100% if this is what causes your problems, but it looks as if it's at least involved in your issue. Either way - I'll get to work on this as soon as I'm done with the first stage of server migration. (Around Feb 13/14.)

**Comment by rsimon** [comment]
P.S.: note that this behavior __really__ just happens after a fresh reload. If you create several annotations on top of each other without reloading the page, there will be multiple highlights on top of each other; and you'll be able to delete them one by one.

**Comment by rsimon** [comment]
P.P.S.: going back to your example above, I think I traced this down in the CSV. There are two annotations

- 091a4459-6922-40f5-a28e-d41244af3bc7, an unverified PLACE, with tags `natural feature` and `isthmus`
- 6f912ab7-8007-431f-a768-0e89376d7228. am EVENT with tags `extent`, `relative`

They are at the same char offset, but only one of them appears after a fresh reload.

**Comment by ChiaraPalladino** [comment]
yeah having no graphical distinction between stratified annotation makes multiple annotations difficult, but I was aware of that. 
My issue here is that the event annotation in that charset is **not** visualized in recogito, that's why I included a screenshot. in fact, both the IDs correspond to the string τὸν ἕως Κασπίας ἰσθμόν, which as you can see is only highlighted in green. my intention was to annotate that string as two different entities, but only one appears in the interface, although the other one is actually saved. 
Generally when multiple entities are annotated you can see the stratification of different colors, see Εὐρώπης a bit below in my screenshot. In the problematic case, there is only green and no purple at all.

**Comment by rsimon** [comment]
Yes, that's exactly what I meant. Only the Place annotation shows, but not the event. Delete the place annotation, and the UI will look as if there is no annotation left. But on the server, the Event annotation is still stored; and you should see it if you refresh the page. 

You _could_ see the stratification in different colours if you made the Event annotation first, and then a Place annotation on top of it, without refreshing the page in between.

**Comment by rsimon** [comment]
Since the original issue from the tile (Event export) is resolved, I'm closing this issue. The bigger part -  annotation stratification - now has its own issue (#467) and was added to the "Annotation UI Enhancements" milestone.

**Comment by rsimon** [comment]
@ChiaraPalladino I'm starting to address this now. A first fix is already in. Creating a selection that perfectly overlaps an existing annotation now no longer creates a new annotation on top of it, but simply opens the (top-most) perfectly overlapping one. 

Before starting this, I played around a bit to understand the browser behavior better. My suspicon is now: the reason that you managed to accumulate so many stratified annotations might be that you double clicked on existing annotations (rather than single-clicked) to open them? When you did that on a single-word annotation, browser behavior interfered, and created a new highlight, thus creating a new annotation. 

That's now resolved & double clicks are now fine (as they should be :-) Undesired stratification should no longer happen. I'm now looking into solutions for visualizing and resolving stratification that already exists.

**Comment by rsimon** [comment]
@ChiaraPalladino another update. I now fixed another thing that, as I just realized, was causing further trouble: stratification would have been less of an issue, I guess, if highlights in the UI would have properly stacked. At least then, you could have deleted the topmost, gotten to the next layer, etc. Also, more stratified highlights would at least have appeared darker. Instead, as we've seen, there ever was only one highlight, and deleting that just removed the highlight completely.

The lastest fix makes sure highlights are properly stacked, meaning you can "dig down", deleting from the top-most to the lowest; and they get darker as they stack.

I'll also look into ways of making the visualization stronger, as it were. Underlines etc., something that makes it more visible than just a slight difference in opacity.

**Comment by rsimon** [comment]
Stratified annotations are now shown with addtional underlines:

![image](https://user-images.githubusercontent.com/470971/39049982-dd8850b8-44a3-11e8-9845-71e01a34282f.png)

**Comment by eltonteb** [comment]
I like this.

**Comment by ChiaraPalladino** [comment]
this looks great Rainer!
I'll have a look at the stratification. I don't double-click to open the annotations, but i guess at a certain point I annotated the same thing multiple times because I wasn't seeing the highlights and I assumed they hadn't been properly saved.

**Comment by rsimon** [comment]
> I annotated the same thing multiple times because I wasn't seeing the highlights and I assumed they hadn't been properly saved.

Ah, yes of course. That would also explain it!
